runtime/server/metrics_views.go

```diff
@@ -78,89 +58,223 @@ func (s *Server) MetricsViewToplist(ctx context.Context, req *api.MetricsViewTop
+		if err != nil {
+			return "", nil, err
+		}
+		whereClause += clause
+		args = append(args, clauseArgs...)
+	}
+
+	if req.Sort != nil {
+		// TODO
+	}
+
+	sql := fmt.Sprintf("SELECT %s FROM %s WHERE %s GROUP BY %s LIMIT %d",
+		strings.Join(selectCols, ", "), req.MetricsViewName, whereClause, req.DimensionName, req.Limit)
+	return sql, args, nil
+}
+
+func buildMetricsTotalsSql(req *api.MetricsViewTotalsRequest) (string, []any, error) {
+	// TODO: get from Catalog
+	timeField := "timestamp"
+	whereClause := fmt.Sprintf("%s >= epoch_ms(?) AND %s < epoch_ms(?) ", timeField, timeField)
+	args := []any{req.TimeStart, req.TimeEnd}
+
+	if req.Filter != nil {
+		clause, clauseArgs, err := buildFilterClauseForMetricsViewFilter(req.Filter)
 		if err != nil {
 			return "", nil, err
 		}
 		whereClause += clause
 		args = append(args, clauseArgs...)
 	}
 
-	sql := fmt.Sprintf("SELECT %s FROM %s WHERE %s GROUP BY %s LIMIT 1000", strings.Join(selectCols, ", "), req.MetricsViewName, whereClause, timeField)
+	sql := fmt.Sprintf("SELECT %s FROM %s WHERE %s",
+		strings.Join(req.MeasureNames, ", "), req.MetricsViewName, whereClause)
 	return sql, args, nil
 }
 
-func buildFilterClause(conds []*api.MetricsViewFilter_Cond, operator string) (string, []any, error) {
-	args := []any{}
+// Builds clause and args for api.MetricsViewFilter
+func buildFilterClauseForMetricsViewFilter(filter *api.MetricsViewFilter) (string, []any, error) {
+	whereClause := ""
+	var args []any
+
+	if filter != nil && filter.Include != nil {
+		clause, clauseArgs, err := buildFilterClauseForConditions(filter.Include, "", "OR")
+		if err != nil {
+			return "", nil, err
+		}
+		whereClause += clause
+		args = append(args, clauseArgs...)
+	}
+
+	if filter != nil && filter.Exclude != nil {
+		clause, clauseArgs, err := buildFilterClauseForConditions(filter.Exclude, "NOT", "AND")
+		if err != nil {
+			return "", nil, err
+		}
+		whereClause += clause
+		args = append(args, clauseArgs...)
+	}
+
+	return whereClause, args, nil
+}
+func buildFilterClauseForConditions(conds []*api.MetricsViewFilter_Cond, operatorPrefix string, conditionJoiner string) (string, []any, error) {
 	clause := ""
+	var args []any
+
 	for _, cond := range conds {
-		questionMarks := strings.Join(repeatString("?", len(cond.Values)), ",")
-		clause += fmt.Sprintf("AND %s %s (%s) ", cond.Name, operator, questionMarks)
-		for _, val := range cond.Values {
+		condClause, condArgs, err := buildFilterClauseForCondition(cond, operatorPrefix, conditionJoiner)
+		if err != nil {
+			return "", nil, fmt.Errorf("filter error: %s", err.Error())
+		}
+		if condClause == "" {
+			continue
+		}
+		clause += condClause
+		args = append(args, condArgs...)
+	}
+
+	return clause, args, nil
+}
+func buildFilterClauseForCondition(cond *api.MetricsViewFilter_Cond, operatorPrefix string, conditionJoiner string) (string, []any, error) {
+	var clauses []string
+	var args []any
+
+	if cond.In != nil && len(cond.In) > 0 {
```

Comment:
You can omit the nil check because `len` returns 0 for nil slices: https://go.dev/play/p/5uO4uXJjO6W